rfcs/widget-restructure-rfc-1.md

```diff
@@ -0,0 +1,667 @@
+
+```rust
+#[derive(Debug, WidgetProps)]
+struct MyButtonProps {
+  // Kayak Props
+  #[props(OnEvent)]
+  event_handler: Option<OnEvent>,
+  #[props(Styles)]
+  styles: Option<Style>,
+  #[props(Focusable)]
+  focusable: bool,
+  
+  // Widget Props
+  text: String,
+  disabled: bool,
+}
+```
+
+##### Implementing Widget
+
+Now with everything defined, we're ready to implement `Widget`.
+
+Firstly, the constructor:
+
+```rust
+impl Widget for MyButton {
+  type Props = MyButtonProps;
+  
+  fn constructor(id: Index, children: Children, mut props: Self::Props) -> dyn Widget<Props=Self::Props> where Self: Sized {
+    if props.disabled {
+      // If disabled, also disable focusability
+      props.focusable = false;
+    }
+
+    Self {
+      my_id: id,
+      my_children: children,
+      my_props: props
+    }
+  }
+  
+}
+```
+
+Then we can implement all the getters and setters:
+
+```rust
+fn get_props(&self) -> &Self::Props {
+  &self.my_props
+}
+
+fn get_props_mut(&self) -> &mut Self::Props {
+  &mut self.my_props
+}
+
+fn get_id(&self) -> Index {
+  self.my_id
+}
+
+fn set_id(&mut self, id: Index) {
+	self.my_id = id;  
+}
+
+fn get_name(&self) -> String {
+  String::from("MyButton")
+}
+```
+
+Lastly, the render method.
+
+To help speed this last part up, another change will be proposed. This change actually targets another addition from this RFC: `KayakContextRef`.
+
+##### Extending KayakContextRef
+
+Normally to render out a widget, we need to do something akin to this:
+
+```rust
+// fragment.rs
+
+fn render(&mut self, context: &mut KayakContext) {
+  let parent_id = self.get_id();
+  let tree = crate::WidgetTree::new();
+  tree.add(parent_id, None);
+
+  if let Some(children) = self.children.take() {
+      children(tree.clone(), Some(parent_id), context);
+  } else {
+      return;
+  }
+
+  // Consume the widget tree taking the inner value
+  let tree = tree.take();
+
+  // Evaluate changes to the tree.
+  let changes = context.widget_manager.tree.diff_children(&tree, parent_id);
+
+  context.widget_manager.tree.merge(&tree, parent_id, changes);
+}
+```
+
+But without access to `KayakContext` we can't quite do this anymore (which is good because it's a lot of boilerplate to write).
+
+To address this and clean up the API a bit more, some more code will be placed in `KayakContextRef`.
+
+```rust
+struct KayakContextRef<'a> {
+  // ...
+  widget_tree: WidgetTree
+}
+
+impl<'a> KayakContextRef<'a> {
+  pub(crate) fn new(context: &mut KayakContext, current_id: Index) {
+    let widget_tree = WidgetTree::new();
+    widget_tree.add(current_id, None);
+   	Self {
+      context,
+      current_id,
+      widget_tree,
+    }
+  }
+  
+  // Pretend `Children` isn't wrapped by `Option<...>`
+  pub fn add_children(&mut self, children: Children) {
+    let id =  self.current_id;
+    children(self.widget_tree.clone(), Some(id), context);
+  }
+
+  pub fn add_widget<W: Widget>(&mut self, widget: W, widget_index: usize) {
```

Comment:
ID here should be the `parent_id` I think in this case its the `self.current_id`, but on a second call to `add_widget` `current_id` might be the `child_widget`'s id. Might be worth resetting `current_id` at the end of this function.

Reply:
So update to widget's ID if `should_rerender` (before rendering the widget)? Then reset back to the original at the end of the method?

Reply:
Actually it looks like I have it setup to create a new `KayakContextRef` before rendering the child (although I forgot to give it `child_id`). So it shouldn't need to be updated, right?

Reply:
Hmm, yeah that should be fine 👍